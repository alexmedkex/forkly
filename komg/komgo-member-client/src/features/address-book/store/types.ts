import { List } from 'immutable'
import { ImmutableMap } from '../../../utils/types'
import { Action } from 'redux'
import { IMember } from '../../members/store/types'

export enum AddressBookActionType {
  GET_COMPANIES_REQUEST = '@@addressBook/GET_COMPANIESS_REQUEST',
  GET_COMPANIES_SUCCESS = '@@addressBook/GET_COMPANIESS_SUCCESS',
  GET_COMPANIES_FAILURE = '@@addressBook/GET_COMPANIESS_FAILURE',
  GET_COMPANY_REQUEST = '@@addressBook/GET_COMPANY_REQUEST',
  GET_COMPANY_SUCCESS = '@@addressBook/GET_COMPANY_SUCCESS',
  GET_COMPANY_FAILURE = '@@addressBook/GET_COMPANY_FAILURE',
  GENERATE_MEMBER_PACKAGE_REQUEST = '@@addressBook/GENERATE_MEMBER_PACKAGE_REQUEST',
  GENERATE_MEMBER_PACKAGE_SUCCESS = '@@addressBook/GENERATE_MEMBER_PACKAGE_SUCCESS',
  GENERATE_MEMBER_PACKAGE_FAILURE = '@@addressBook/GENERATE_MEMBER_PACKAGE_FAILURE',
  ADD_COMPANY_TO_ENS_REQUEST = '@@addressBook/ADD_COMPANY_TO_ENS_REQUEST',
  ADD_COMPANY_TO_ENS_SUCCESS = '@@addressBook/ADD_COMPANY_TO_ENS_SUCCESS',
  ADD_COMPANY_TO_ENS_FAILURE = '@@addressBook/ADD_COMPANY_TO_ENS_FAILURE',
  CONFIGURE_MQ_REQUEST = '@@addressBook/CONFIGURE_MQ_REQUEST',
  CONFIGURE_MQ_SUCCESS = '@@addressBook/CONFIGURE_MQ_SUCCESS',
  CONFIGURE_MQ_FAILURE = '@@addressBook/CONFIGURE_MQ_FAILURE',
  CREATE_COMPANY_REQUEST = '@@addressBook/CREATE_COMPANY_REQUEST',
  CREATE_COMPANY_SUCCESS = '@@addressBook/CREATE_COMPANY_SUCCESS',
  CREATE_COMPANY_FAILURE = '@@addressBook/CREATE_COMPANY_FAILURE',
  UPDATE_COMPANY_REQUEST = '@@addressBook/UPDATE_COMPANY_REQUEST',
  UPDATE_COMPANY_SUCCESS = '@@addressBook/UPDATE_COMPANY_SUCCESS',
  UPDATE_COMPANY_FAILURE = '@@addressBook/UPDATE_COMPANY_FAILURE',
  DeleteCompanyByIdSuccess = '@@addressBook/DELETE_COMPANY_BY_ID_SUCCESS'
}

export interface AddressBookStateProperties {
  companies: List<IMember>
}

export type AddressBookState = ImmutableMap<AddressBookStateProperties>

export interface GetCompaniesRequest extends Action {
  type: AddressBookActionType.GET_COMPANIES_SUCCESS
  payload: IMember[]
}
export interface GetCompanyRequest extends Action {
  type: AddressBookActionType.GET_COMPANY_SUCCESS
  payload: IMember
}
export interface GenerateMemberPackageRequest extends Action {
  type: AddressBookActionType.GENERATE_MEMBER_PACKAGE_SUCCESS
  payload: IMember
}
export interface AddCompanyToENSRequest extends Action {
  type: AddressBookActionType.ADD_COMPANY_TO_ENS_SUCCESS
  payload: IMember
}
export interface ConfigureMQRequest extends Action {
  type: AddressBookActionType.CONFIGURE_MQ_SUCCESS
  payload: IMember
}

export interface CreateCompanyRequest extends Action {
  type: AddressBookActionType.CREATE_COMPANY_SUCCESS
  payload: IMember
}
export interface UpdateCompanyRequest extends Action {
  type: AddressBookActionType.UPDATE_COMPANY_SUCCESS
  payload: IMember
}

export interface DeleteCompanyByIdSuccess extends Action {
  type: AddressBookActionType.DeleteCompanyByIdSuccess
  payload: { id: string }
}

export type AddressBookActions =
  | GetCompaniesRequest
  | GetCompanyRequest
  | CreateCompanyRequest
  | UpdateCompanyRequest
  | GenerateMemberPackageRequest
  | AddCompanyToENSRequest
  | ConfigureMQRequest
  | DeleteCompanyByIdSuccess
