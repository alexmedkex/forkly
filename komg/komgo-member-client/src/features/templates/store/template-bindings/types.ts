import { ITemplateBinding } from '@komgo/types'
import { Map, List } from 'immutable'
import { ImmutableMap } from '../../../../utils/types'
import { Action } from 'redux'

export enum EditorTemplateBindingsActionType {
  GET_TEMPLATE_BINDING_REQUEST = '@@template-bindings/GET_TEMPLATE_BINDING_REQUEST',
  GET_TEMPLATE_BINDING_SUCCESS = '@@template-bindings/GET_TEMPLATE_BINDING_SUCCESS',
  GET_TEMPLATE_BINDING_FAILURE = '@@template-bindings/GET_TEMPLATE_BINDING_FAILURE',
  FETCH_TEMPLATE_BINDINGS_REQUEST = '@@template-bindings/FETCH_TEMPLATE_BINDINGS_REQUEST',
  FETCH_TEMPLATE_BINDINGS_SUCCESS = '@@template-bindings/FETCH_TEMPLATE_BINDINGS_SUCCESS',
  FETCH_TEMPLATE_BINDINGS_FAILURE = '@@template-bindings/FETCH_TEMPLATE_BINDINGS_FAILURE'
}

export interface GetTemplateBindingsSuccessAction extends Action {
  type: EditorTemplateBindingsActionType.GET_TEMPLATE_BINDING_SUCCESS
  payload: ITemplateBinding
}

export interface FetchTemplateBindingsSuccessAction extends Action {
  type: EditorTemplateBindingsActionType.FETCH_TEMPLATE_BINDINGS_SUCCESS
  payload: { limit: number; skip: number; total: number; items: ITemplateBinding[] }
}

export interface TemplateBindingsStateProperties {
  byStaticId: Map<string, Map<keyof ITemplateBinding, any>>
  staticIds: List<string>
  total: number
}

export type EditorTemplateBindingsState = ImmutableMap<TemplateBindingsStateProperties>
