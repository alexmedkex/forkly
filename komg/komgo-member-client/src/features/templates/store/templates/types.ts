import { ITemplate } from '@komgo/types'
import { Map, List } from 'immutable'
import { ImmutableMap } from '../../../../utils/types'
import { Action } from 'redux'

export enum EditorTemplatesActionType {
  FETCH_TEMPLATES_REQUEST = '@@templates/FETCH_TEMPLATES_REQUEST',
  FETCH_TEMPLATES_SUCCESS = '@@templates/FETCH_TEMPLATES_SUCCESS',
  FETCH_TEMPLATES_FAILURE = '@@templates/FETCH_TEMPLATES_FAILURE',
  CREATE_TEMPLATE_REQUEST = '@@templates/CREATE_TEMPLATE_REQUEST',
  CREATE_TEMPLATE_SUCCESS = '@@templates/CREATE_TEMPLATE_SUCCESS',
  CREATE_TEMPLATE_FAILURE = '@@templates/CREATE_TEMPLATE_FAILURE',
  UPDATE_TEMPLATE_REQUEST = '@@templates/UPDATE_TEMPLATE_REQUEST',
  UPDATE_TEMPLATE_SUCCESS = '@@templates/UPDATE_TEMPLATE_SUCCESS',
  UPDATE_TEMPLATE_FAILURE = '@@templates/UPDATE_TEMPLATE_FAILURE',
  GET_TEMPLATE_REQUEST = '@@templates/GET_TEMPLATE_REQUEST',
  GET_TEMPLATE_SUCCESS = '@@templates/GET_TEMPLATE_SUCCESS',
  GET_TEMPLATE_FAILURE = '@@templates/GET_TEMPLATE_FAILURE',
  DELETE_TEMPLATE_REQUEST = '@@templates/DELETE_TEMPLATE_REQUEST',
  DELETE_TEMPLATE_SUCCESS = '@@templates/DELETE_TEMPLATE_SUCCESS',
  DELETE_TEMPLATE_FAILURE = '@@templates/DELETE_TEMPLATE_FAILURE'
}

export interface FetchTemplatesSuccessAction extends Action {
  type: EditorTemplatesActionType.FETCH_TEMPLATES_SUCCESS
  payload: { limit: number; skip: number; total: number; items: ITemplate[] }
}

export interface TemplateStateProperties {
  byStaticId: Map<string, Map<keyof ITemplate, any>>
  staticIds: List<string>
  total: number
}

export type EditorTemplatesState = ImmutableMap<TemplateStateProperties>
