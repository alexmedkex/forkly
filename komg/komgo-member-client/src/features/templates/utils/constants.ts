export const EMPTY_TEMPLATE = {
  object: 'value',
  document: {
    object: 'document',
    data: {},
    nodes: [
      {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [
          {
            object: 'text',
            text: 'Empty Template',
            marks: [{ object: 'mark', type: 'bold', data: {} }]
          }
        ]
      },
      {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [
          {
            object: 'text',
            text: '',
            marks: [{ object: 'mark', type: 'bold', data: {} }]
          }
        ]
      },
      {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [
          {
            object: 'text',
            text: 'Please add your wording then add the dynamic fields you need from the above',
            marks: []
          },
          {
            object: 'text',
            text: ' Add Field',
            marks: [{ object: 'mark', type: 'bold', data: {} }]
          },
          { object: 'text', text: ' button', marks: [] }
        ]
      },
      {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [{ object: 'text', text: '', marks: [] }]
      },
      {
        object: 'block',
        type: 'paragraph',
        data: {},
        nodes: [
          {
            object: 'text',
            text: 'e.g. Some text then a variable: ',
            marks: []
          },
          {
            object: 'inline',
            type: 'variable',
            data: { path: 'applicant.x500Name.CN', title: 'Applicant name' },
            nodes: [{ object: 'text', text: 'Applicant name', marks: [] }]
          },
          { object: 'text', text: '', marks: [] }
        ]
      }
    ]
  }
}
