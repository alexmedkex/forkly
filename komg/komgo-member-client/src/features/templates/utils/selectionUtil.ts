import { parse } from 'qs'

export interface ISelection {
  select: boolean
  redirectTo: string
  type: string // TODO LS we need to decide how to mark the template. I don't think LetterOfCreditType is the right one
}

const REDIRECT_WHITE_LIST = [/^\/letters-of-credit\/new/]

export const buildSelection = (querystring: string): ISelection => {
  const { select = false, type, redirectTo } = parse(querystring, {
    ignoreQueryPrefix: true,
    decoder: str => {
      const keywords = {
        true: true,
        false: false,
        null: null,
        undefined
      }
      if (str in keywords) {
        return keywords[str]
      }
      return decodeURIComponent(str)
    }
  })

  if (redirectTo && !REDIRECT_WHITE_LIST.some(route => redirectTo.match(route))) {
    throw new Error(`Redirect to ${redirectTo} not supported`)
  }

  return {
    redirectTo,
    select,
    type
  }
}
