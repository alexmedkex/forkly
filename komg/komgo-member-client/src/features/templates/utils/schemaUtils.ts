import { findLetterOfCreditSchema } from '../../letter-of-credit/utils/findLetterOfCreditSchema'

export const resolveBindings = (
  bindings: { [key: string]: string },
  schemaResolver: (key: string) => any = findLetterOfCreditSchema
): { [key: string]: any } =>
  Object.entries(bindings).reduce(
    (memo, [key, value]) => ({
      ...memo,
      [key]: schemaResolver(value)
    }),
    {}
  )

export const findTemplateBindingType = (schemaId: string) => {
  // FIXME LS until we don't add a templateBiding name
  return schemaId.match(/http:\/\/komgo.io\/schema\/(.*)\/template-bindings/)[1].toUpperCase()
}
