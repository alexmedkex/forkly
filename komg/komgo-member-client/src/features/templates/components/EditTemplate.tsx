import React, { SyntheticEvent, useState } from 'react'
import { Button, Form, Header, Input } from 'semantic-ui-react'
import { Preview } from '../components/Preview'
import { ViewContainer } from '../components/ViewContainer'
import { SPACES, Template, TemplateMode } from '@komgo/ui-components'
import { buildFields, ITemplate, ITemplateBinding } from '@komgo/types'
import { findTemplateBindingType, resolveBindings } from '../utils/schemaUtils'
import { findLetterOfCreditSchema } from '../../letter-of-credit/utils/findLetterOfCreditSchema'
import { Formik, FormikProps } from 'formik'
import { ServerError } from '../../../store/common/types'
import { RouteComponentProps } from 'react-router'
import { ImmutableObject } from '../../../utils/types'
import { FILTERED_FIELDS } from '../../letter-of-credit/constants'
import { TopHeader } from './TopHeader'
import { ISelection } from '../utils/selectionUtil'
import { Spacer } from '../../../components/spacer/Spacer'

export interface EditTemplateProps extends RouteComponentProps<any> {
  selection?: ISelection
  template: ImmutableObject<ITemplate>
  templateBinding: ImmutableObject<ITemplateBinding>
  updateTemplate: (template: ITemplate) => void
  getTemplateWithTemplateBindings: (params: { staticId: string }) => void
  isUpdating: boolean
  updatingErrors: ServerError[]
}

export const EditTemplate = ({
  templateBinding,
  template,
  updateTemplate,
  isUpdating,
  updatingErrors,
  history,
  selection = {} as any
}: EditTemplateProps) => {
  // Cache toJS result
  const [templateModel] = useState<ITemplate>(() => template.toJS())
  const [templateBindingModel] = useState<ITemplateBinding>(() => templateBinding.toJS())

  const { bindings, dataSchemaId, templateSchemaId, example } = templateBindingModel
  // FIXME LS move buildFields as internal of Template
  // New TemplateOptions:
  // - schemas: map or array of schemas => contains all the schemas used by the template
  // - dataSchema: string
  // - templateSchema: string

  const fields = bindings
    ? buildFields({
        bindings: resolveBindings(bindings, findLetterOfCreditSchema),
        dataSchema: findLetterOfCreditSchema(dataSchemaId),
        templateSchema: findLetterOfCreditSchema(templateSchemaId)
      }).filter(f => !FILTERED_FIELDS.includes(f.dataPath))
    : []

  const [error] = updatingErrors

  const type = findTemplateBindingType(templateSchemaId) // e.g. SBLC / LC

  const onSelectClick = (template: ImmutableObject<ITemplate>, e: SyntheticEvent) => {
    e.preventDefault()
    e.stopPropagation()
    history.push(`${selection.redirectTo}&templateId=${template.get('staticId')}`)
  }

  return (
    <>
      <Formik
        onSubmit={(values, actions) => {
          updateTemplate(values)
          actions.resetForm(values)
        }}
        initialValues={templateModel}
        render={({ values, handleSubmit, setFieldValue, setFieldTouched, dirty }: FormikProps<ITemplate>) => {
          const onTemplateChange = (value, mode) => {
            // FIXME the template should return the immutable object
            if (template.get('template') !== value) {
              setFieldValue('template', value)
              setFieldTouched('template', true)
            }
          }

          const onNameChange = (_, { name, value }) => {
            setFieldValue(name, value)
            setFieldTouched(name, true)
          }

          return (
            <Form onSubmit={handleSubmit}>
              <TopHeader>
                <Header as="h1" style={{ margin: 0 }}>
                  {/* TODO LS keep until we decide <Breadcrumb as="small">Financial Instruments / Templates</Breadcrumb>*/}
                  {selection.select ? (
                    `Templates`
                  ) : (
                    <Form.Field inline={true} style={{ fontSize: '14px' }}>
                      <label htmlFor="field_name">{type}:</label>
                      <Input
                        id="field_name"
                        onChange={onNameChange}
                        type="text"
                        style={{ width: '300px' }}
                        name="name"
                        value={values.name}
                      />
                    </Form.Field>
                  )}
                </Header>

                {selection.select ? (
                  <Button
                    data-test-id="select-template"
                    type="button"
                    onClick={onSelectClick.bind(undefined, template)}
                  >
                    Select
                  </Button>
                ) : (
                  <Button
                    data-test-id="save-template"
                    type="submit"
                    disabled={!dirty && !error}
                    primary={true}
                    loading={isUpdating}
                  >
                    Save
                  </Button>
                )}
              </TopHeader>

              <Spacer marginBottom={SPACES.DEFAULT} paddingRight={SPACES.DEFAULT} paddingLeft={SPACES.DEFAULT}>
                {selection.select && (
                  <p>
                    Choose <strong>{values.name}</strong> template to begin your application
                  </p>
                )}
              </Spacer>

              <ViewContainer>
                <Preview height={'104px'}>
                  {/*
                  // FIXME LS we need to dig into this one on the Template. It happens when we leave the page
                  // Warning: Can't perform a React state update on an unmounted component.
                  // This is a no-op, but it indicates a memory leak in your application.
                  // To fix, cancel all subscriptions and asynchronous tasks in a useEffect cleanup function.
                  */}
                  <Template
                    modeLocked={selection.select}
                    mode={selection.select ? TemplateMode.ReadOnly : TemplateMode.Write}
                    ownerCompanyStaticId={template.get('ownerCompanyStaticId')}
                    fields={fields}
                    data={example}
                    template={templateModel.template}
                    onChange={onTemplateChange}
                  />
                </Preview>
              </ViewContainer>
            </Form>
          )
        }}
      />
    </>
  )
}
