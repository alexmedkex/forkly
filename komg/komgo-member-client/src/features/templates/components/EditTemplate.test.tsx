import * as React from 'react'
import { render, cleanup, fireEvent, act, wait } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { EditTemplate, EditTemplateProps } from './EditTemplate'
import { MemoryRouter as Router } from 'react-router'
import { buildFakeTemplate, buildFakeTemplateBinding } from '@komgo/types'
import { fromJS } from 'immutable'
import { buildSelection } from '../utils/selectionUtil'

const globalAsAny = global as any

const originalGetSelection = globalAsAny.window.getSelection

describe('EditTemplate', () => {
  let props: EditTemplateProps

  beforeEach(() => {
    globalAsAny.window.getSelection = jest.fn(() => {
      return {}
    })

    props = {
      template: fromJS(buildFakeTemplate()),
      updateTemplate: jest.fn(),
      getTemplateWithTemplateBindings: jest.fn(),
      isUpdating: false,
      updatingErrors: [],
      templateBinding: fromJS(buildFakeTemplateBinding()),
      history: { push: jest.fn() } as any,
      location: {
        pathname: '',
        search: '',
        state: '',
        hash: ''
      },
      match: {
        isExact: true,
        path: '',
        url: '',
        params: null
      },
      staticContext: undefined
    }
  })

  afterEach(() => {
    globalAsAny.window.getSelection = originalGetSelection
    afterEach(cleanup)
  })

  describe('view', () => {
    it('renders', () => {
      act(() => {
        const { asFragment } = render(
          <Router>
            <EditTemplate {...props} />
          </Router>
        )
        expect(asFragment()).toMatchSnapshot()
      })
    })

    describe('save', () => {
      it('saves the template changes', async () => {
        const newName = 'A new name'
        let component

        act(() => {
          component = render(
            <Router>
              <EditTemplate {...props} />
            </Router>
          )
        })

        act(() => {
          userEvent.type(component.getByLabelText('SBLC:'), newName)
          fireEvent.click(component.getByText('Save'))
        })

        await wait(() => {
          act(() => {
            expect(props.updateTemplate).toHaveBeenCalledWith({
              ...props.template.toJS(),
              name: newName
            })
          })
        })
      })
    })
  })

  describe('selection', () => {
    it('renders', () => {
      const selection = buildSelection(
        '?select=true&redirectTo=%2Fletters-of-credit%2Fnew%3Fsource%3DKOMGO%26sourceId%3Dbf6e713b-53be-4510-9be5-bf47f2171d6b&type=SBLC'
      )
      const { asFragment } = render(
        <Router>
          <EditTemplate {...props} selection={selection} />
        </Router>
      )
      expect(asFragment()).toMatchSnapshot()
    })

    describe('select', () => {
      it('redirect to the sblc', async () => {
        const selection = buildSelection(
          '?select=true&redirectTo=%2Fletters-of-credit%2Fnew%3Fsource%3DKOMGO%26sourceId%3Dbf6e713b-53be-4510-9be5-bf47f2171d6b&type=SBLC'
        )

        let component

        act(() => {
          component = render(
            <Router>
              <EditTemplate {...props} selection={selection} />
            </Router>
          )
        })

        act(() => {
          fireEvent.click(component.getByText('Select'))
        })

        expect(props.history.push).toHaveBeenCalledWith(
          `${selection.redirectTo}&templateId=${props.template.get('staticId')}`
        )
      })
    })
  })
})
