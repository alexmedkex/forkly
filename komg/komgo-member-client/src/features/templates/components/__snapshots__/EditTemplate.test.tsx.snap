// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditTemplate selection renders 1`] = `
<DocumentFragment>
  .c8 {
  background: rgba(148,109,192,0.4);
  font-style: normal;
  -webkit-text-decoration: none;
  text-decoration: none;
  font-weight: normal;
  text-transform: uppercase;
}

.c5 {
  width: 100%;
  padding: 8px;
  min-height: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-shadow: 0 4px 2px -2px rgba(192,207,222,0.51);
}

.c6 {
  margin-left: auto;
}

.c7 {
  margin-top: 2px;
  padding: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 1px solid #c0cfde;
  background-color: white;
  box-shadow: 0 4px 2px -2px rgba(192,207,222,0.51);
  height: 100%;
  width: 100%;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  background-color: #f2f5f8;
}

.c2 {
  margin-top: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
  width: 100%;
  border-top: 1px solid #c0cfde;
  background-color: #f2f5f8;
}

.c0 {
  padding: 0;
  margin: 0;
  padding-top: 16px;
  padding-bottom: 24px;
  padding-left: 24px;
  padding-right: 24px;
}

.c1 {
  padding: 0;
  margin: 0;
  padding-left: 24px;
  padding-right: 24px;
  margin-bottom: 24px;
}

@media (min-width:68.75em) {
  .c3 {
    max-width: 990px;
    margin-left: auto;
    margin-right: auto;
    padding: 16px;
    height: calc(100vh - 104px);
  }
}

@media (min-width:68.75em) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    height: 100%;
    overflow: none;
  }
}

<form
    class="ui form"
  >
    <div
      class="c0"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: flex-start;"
      >
        <h1
          class="ui header"
          style="margin: 0px;"
        >
          Templates
        </h1>
        <button
          class="ui button"
          data-test-id="select-template"
          role="button"
          type="button"
        >
          Select
        </button>
      </div>
    </div>
    <div
      class="c1"
    >
      <p>
        Choose 
        <strong>
          FakeTemplate
        </strong>
         template to begin your application
      </p>
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
        height="104px"
      >
        <div
          class="c4"
        >
          <div
            class="c5"
          >
            <div
              style="padding-left: 16px; color: rgb(212, 92, 100);"
            >
              <i
                aria-hidden="true"
                class="red warning sign icon"
              />
              <span
                style="padding-left: 16px;"
              >
                The template is missing 8 mandatory fields. 
              </span>
            </div>
            <div
              class="c6"
            >
              <div
                style="display: flex; align-items: center;"
              >
                <div
                  class="ui disabled fitted toggle checkbox"
                  data-test-id="button-mode"
                >
                  <input
                    class="hidden"
                    disabled=""
                    readonly=""
                    tabindex="-1"
                    type="checkbox"
                    value=""
                  />
                  <label />
                </div>
                <label
                  style="padding-right: 16px; padding-left: 16px;"
                >
                  Edit mode
                </label>
              </div>
            </div>
          </div>
          <div
            class="c7"
          >
            <div
              autocorrect="on"
              data-gramm="false"
              data-key="124"
              data-slate-editor="true"
              spellcheck="true"
              style="outline: none; white-space: pre-wrap; word-wrap: break-word; width: 100%; height: 100%; flex-grow: 1;"
            >
              <div
                data-key="125"
                data-slate-object="block"
                style="position: relative;"
              >
                <span
                  data-key="126"
                  data-slate-object="text"
                >
                  <span
                    data-offset-key="126:0"
                    data-slate-leaf="true"
                  >
                    <span
                      data-slate-string="true"
                    >
                      some text 
                    </span>
                  </span>
                </span>
                <span
                  data-key="127"
                  data-slate-void="true"
                >
                  <span>
                    <var
                      class="c8"
                      data-key="127"
                      data-slate-object="inline"
                      data-test-id="var-trade.quantity"
                    >
                      600000
                    </var>
                  </span>
                </span>
                <span
                  data-key="129"
                  data-slate-object="text"
                >
                  <span
                    data-offset-key="129:0"
                    data-slate-leaf="true"
                  >
                    <span
                      data-slate-string="true"
                    >
                       
                    </span>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;

exports[`EditTemplate view renders 1`] = `
<DocumentFragment>
  .c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  border: 1px solid #c0cfde;
  background-color: white;
  box-shadow: 0 4px 2px -2px rgba(192,207,222,0.51);
  height: 100%;
  width: 100%;
}

.c3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  background-color: #f2f5f8;
}

.c2 {
  margin-top: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: auto;
  width: 100%;
  border-top: 1px solid #c0cfde;
  background-color: #f2f5f8;
}

.c0 {
  padding: 0;
  margin: 0;
  padding-top: 16px;
  padding-bottom: 24px;
  padding-left: 24px;
  padding-right: 24px;
}

.c1 {
  padding: 0;
  margin: 0;
  padding-left: 24px;
  padding-right: 24px;
  margin-bottom: 24px;
}

@media (min-width:68.75em) {
  .c3 {
    max-width: 990px;
    margin-left: auto;
    margin-right: auto;
    padding: 16px;
    height: calc(100vh - 104px);
  }
}

@media (min-width:68.75em) {
  .c2 {
    -webkit-flex-direction: row;
    -ms-flex-direction: row;
    flex-direction: row;
    height: 100%;
    overflow: none;
  }
}

<form
    class="ui form"
  >
    <div
      class="c0"
    >
      <div
        style="display: flex; justify-content: space-between; align-items: flex-start;"
      >
        <h1
          class="ui header"
          style="margin: 0px;"
        >
          <div
            class="inline field"
            style="font-size: 14px;"
          >
            <label
              for="field_name"
            >
              SBLC:
            </label>
            <div
              class="ui input"
              style="width: 300px;"
            >
              <input
                id="field_name"
                name="name"
                type="text"
                value="FakeTemplate"
              />
            </div>
          </div>
        </h1>
        <button
          class="ui primary disabled button"
          data-test-id="save-template"
          disabled=""
          role="button"
          tabindex="-1"
          type="submit"
        >
          Save
        </button>
      </div>
    </div>
    <div
      class="c1"
    />
    <div
      class="c2"
    >
      <div
        class="c3"
        height="104px"
      >
        <div
          class="c4"
        >
          <div
            autocorrect="on"
            contenteditable="true"
            data-gramm="false"
            data-key="0"
            data-slate-editor="true"
            role="textbox"
            spellcheck="true"
            style="outline: none; white-space: pre-wrap; word-wrap: break-word; width: 100%; height: 100%; flex-grow: 1;"
          >
            <div
              data-key="1"
              data-slate-object="block"
              style="position: relative;"
            >
              <span
                data-key="2"
                data-slate-object="text"
              >
                <span
                  data-offset-key="2:0"
                  data-slate-leaf="true"
                >
                  <span
                    data-slate-string="true"
                  >
                    some text 
                  </span>
                </span>
              </span>
              <span
                contenteditable="false"
                data-key="3"
                data-slate-void="true"
              >
                <span
                  data-slate-spacer="true"
                  style="height: 0px; color: transparent; outline: none; position: absolute;"
                >
                  <span
                    data-key="4"
                    data-slate-object="text"
                  >
                    <span
                      data-offset-key="4:0"
                      data-slate-leaf="true"
                    >
                      <span
                        data-slate-length="14"
                        data-slate-zero-width="z"
                      >
                        ﻿
                      </span>
                    </span>
                  </span>
                </span>
                <span
                  contenteditable="false"
                >
                  <span
                    data-key="3"
                    data-slate-object="inline"
                    style="position: relative;"
                  >
                    <span
                      data-key="4"
                      data-slate-object="text"
                    >
                      <span
                        data-offset-key="4:0"
                        data-slate-leaf="true"
                      >
                        <span
                          data-slate-length="14"
                          data-slate-zero-width="z"
                        >
                          ﻿
                        </span>
                      </span>
                    </span>
                  </span>
                </span>
              </span>
              <span
                data-key="5"
                data-slate-object="text"
              >
                <span
                  data-offset-key="5:0"
                  data-slate-leaf="true"
                >
                  <span
                    data-slate-string="true"
                  >
                     
                  </span>
                </span>
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
